src/stm.ts

```diff
@@ -0,0 +1,515 @@
+   * return the same data.
+   */
+  RepeatableReads,
+
+  /**
+   * ReadCommitted reads keys from any committed revision.
+   */
+  ReadCommitted,
+}
+/**
+ * ISTMOptions are optionally passed to `etcd3.stm(options)`.
+ */
+export interface ISTMOptions {
+  /**
+   * Number of times we'll retry the transaction if we get a conflict.
+   * Defaults to 3.
+   */
+  retries: number;
+
+  /**
+   * WithPrefetch is a hint to prefetch a list of keys before trying to apply.
+   * If an STM transaction will unconditionally fetch a set of keys, prefetching
+   * those keys will save the round-trip cost from requesting
+   * each key one by one with `.get()`.
+   */
+  prefetch: string[];
+
+  /**
+   * Isolation level for the transaction. Defaults to SerializableSnapshot.
+   */
+  isolation: Isolation;
+}
+
+/**
+ * Converts the key/value pair to a partial response that contains it. The
+ * response *will not* contain header or revision information.
+ */
+function keyValueToResponse(key: string | Buffer, value?: Buffer): RPC.IRangeResponse {
+  key = toBuffer(key);
+
+  if (!value) {
+    return <any>{ kvs: [], more: false, count: '0' };
+  }
+
+  return <any>{
+    kvs: [
+      {
+        key: Buffer.from(key),
+        value,
+      },
+    ],
+    more: false,
+    count: '1',
+  };
+}
+
+/**
+ * ReadSet records a set of reads in a SoftwareTransaction.
+ */
+class ReadSet {
+  private readonly reads: { [key: string]: Promise<RPC.IRangeResponse> } = Object.create(null);
+  private readonly completedReads: { key: Buffer; res: RPC.IRangeResponse }[] = [];
+  private earliestMod = new BigNumber(Infinity);
+
+  /**
+   * Returns the earliest modified revision of any key in this change set.
+   */
+  public earliestModRevision(): BigNumber {
+    return this.earliestMod;
+  }
+
+  /**
+   * Add checks to the comparator to make sure that the mod revision of all
+   * keys read during the transaction are the same.
+   */
+  public addCurrentChecks(cmp: Builder.ComparatorBuilder) {
+    this.completedReads.forEach(({ key, res }) => {
+      if (res.kvs.length) {
+        cmp.and(key, 'Mod', '==', res.kvs[0].mod_revision);
+      } else {
+        cmp.and(key, 'Mod', '==', 0);
+      }
+    });
+  }
+
+  /**
+   * injectKeyValue manually inserts the value as the find response
+   * when looking up the given key.
+   */
+  public injectKeyValue(key: string, value?: Buffer) {
+    this.reads[key] = Promise.resolve(keyValueToResponse(key, value));
+  }
+
+  /**
+   * runRequest sets read options and executes the outgoing request.
+   */
+  public runRequest(kv: RPC.KVClient, req: RPC.IRangeRequest): Promise<RPC.IRangeResponse> {
+    const key = req.key!.toString();
+    if (this.reads[key]) {
+      return this.reads[key];
+    }
+
+    const promise = kv.range(req).then(res => {
+      this.earliestMod = BigNumber.min(new BigNumber(res.kvs[0].mod_revision), this.earliestMod);
+      this.completedReads.push({ key: req.key!, res });
```

Comment:
You can use `!` as a suffix?

Reply:
Yep! It's a non-null assertion. In the protobuf-generated interfaces for requests, all properties are marked as optional, as we can't statically determine which ones are and aren't. But this key is actually always required and set by the RangeBuilder, we assert that here.